src/js/glob.js

```diff
@@ -0,0 +1,349 @@
+ * @param {String} pattern - A single "glob" pattern.
+ * @param {Object} [minimatchOptions] - (Optional) options to pass to minimatch.
+ * @param {Boolean} [isDir] - (Optional) Whether or not the path refers to a directory.
+ * @return {Boolean} `true` if the pattern matches, `false` if not.
+ *
+ */
+gpii.glob.matchesSinglePattern = function (pathToMatch, pattern, minimatchOptions, isDir) {
+    minimatchOptions = minimatchOptions || {};
+
+    if (isDir) {
+        return gpii.glob.dirMightMatch(pathToMatch, pattern);
+    }
+    else {
+        return minimatch(pathToMatch, pattern, minimatchOptions);
+    }
+};
+
+/**
+ *
+ * Match a directory against a pattern and return true if it might contain material that matches the pattern.
+ *
+ * @param {String} pathToDir - The full path to the directory.
+ * @param {String} pattern - The (positive) pattern to test the path against.
+ * @return {Boolean} `true` if the directory might contain matches, `false` otherwise.
+ *
+ */
+gpii.glob.dirMightMatch = function (pathToDir, pattern) {
+    // We use the equivalent of the basePath option in minimatch, i.e. any directory might contain a pattern with no slashes.
+    if (pattern.indexOf("/") !== -1) {
+        var patternSegments = pattern.split("/");
+        var pathSegments    = pathToDir.split(path.posix.sep);
+
+        for (var a = 0; a < pathSegments.length; a++) {
+            var patternSegment = patternSegments[a];
+
+            // If we make it to a directory wildcard, there may be matches in the dir or one of its children.
+            if (patternSegment === "**") {
+                return true;
+            }
+
+            var pathSegment = pathSegments[a];
+            if (pathSegment !== patternSegment) {
+                return false;
+            }
+        }
+    }
+
+    return true;
+};
+
+// The default list of regular expressions that describe "invalid globs".
+gpii.glob.invalidGlobRules = {
+    noLeadingWildcard: /^(\.\/)?\*\*/,
+    noWindowsSeparator: /\\/,
+    noParentDir: /^\.\./,
+    noRegexp: /[\[\](){}|]/,
+    noWholeRoot: /^\.\/$/
+};
+
+/**
+ *
+ * Check a pattern to ensure that it conforms to our constraints, which are:
+ *
+ * 1. It must not contain a leading wildcard, as in "**" or "./**".
+ * 2. It must not contain windows-style separators, i.e. backslashes.
+ * 3. It must not begin with a "parent" operator, i.e. "../"
+ *
+ * @param {String} pattern - A pattern to evaluate.
+ * @param {Object|Array<String>} [rules] - An optional set of custom rules defining invalid patterns as regular expressions.
+ * @return {Boolean} `true` if the pattern is valid, `false` otherwise.
+ *
+ */
+gpii.glob.isValidPattern = function (pattern, rules) {
+    var positivePattern = gpii.glob.positivePattern(pattern);
+    rules = rules || gpii.glob.invalidGlobRules;
+
+    var failsAtLeastOneRule = false;
+    fluid.each(rules, function (invalidGlobRule) {
+        if (positivePattern.match(invalidGlobRule)) {
```

Comment:
This validator should supply better diagnostics, and give the user some clue about why the pattern is invalid

Reply:
I expanded the patterns into a pattern and message construct so that it's clear what the exact failure is in all cases.